### ISSUE TITLE
12-nlp-url-spam.ipynb

### ISSUE THREAD

**Issue by mezcolantriz** [4GeeksAcademy/elius123ef#8]
Hola @elius123ef , por aquí las issues del ejercicio de NLP!

## LO QUE HAS HECHO BIEN

- **Importaciones completas**: pandas, numpy, re, matplotlib, nltk, pickle, TfidfVectorizer, train_test_split, SVC, GridSearchCV, RandomizedSearchCV
- **Limpieza de datos**: duplicados eliminados, conversión de tipos
- **Preprocesamiento de texto completo**:
  - Eliminación de `https://`
  - Tokenización
  - Lowercase
  - Lematización con WordNetLemmatizer
  - Eliminación de stopwords
  - Filtrado por longitud (>3)
- **WordCloud** para visualización
- **TfidfVectorizer bien configurado**: `max_features=5000`, `max_df=0.8`, `min_df=5`
- **Train/test split con stratify** (muy bien!)
- **Modelo SVC inicial** con kernel lineal
- **Métricas de train y test** para detectar overfitting
- **GridSearchCV Y RandomizedSearchCV** - probaste ambos 
- **Matriz de confusión visualizada** con `ConfusionMatrixDisplay`
- **Modelo guardado con pickle** 

### 1. **Variable `svc` no definida en RandomizedSearchCV**
```python
# Usas 'svc' pero nunca la definiste
random_search = RandomizedSearchCV(
    estimator=svc,  # ← ERROR: 'svc' no existe ¿código eliminado?
    ...
)

# Deberías usar 'model' o crear una nueva instancia:
random_search = RandomizedSearchCV(
    estimator=SVC(random_state=18),
    ...
)
```

### 2. **Falta guardar el vectorizer (IMPORTANTE)**
```python
# Solo guardas el modelo
with open('../models/12-nlp-url-spam-best_model.pkl', 'wb') as f:
    pickle.dump(best_model, f)

# También debes guardar el vectorizer:
with open('../models/12-nlp-url-spam-vectorizer.pkl', 'wb') as f:
    pickle.dump(vectorizer, f)
```
**Sin el vectorizer no podrás hacer predicciones con URLs nuevas.**

### 3. **param_dist incorrecto para RandomizedSearchCV**
```python
# C como tupla no funciona como distribución continua
param_dist = {'C': (0.1, 100),  # ← Esto no es una distribución
              ...}

# Usa una distribución real o lista:
from scipy.stats import uniform
param_dist = {'C': uniform(loc=0.1, scale=100),  # Distribución uniforme
              'kernel': ['linear', 'rbf'],
              'gamma': ['scale', 'auto']}

# O simplemente usa una lista como en GridSearch:
param_dist = {'C': [0.1, 1, 10, 50, 100],
              ...}
```

### 4. **Faltan más métricas además de accuracy**
```python
#  Solo muestras accuracy
metrics = {"Accuracy Train: ": ..., "Accuracy Test ": ...}

# Para clasificación binaria añade:

classification_report(y_test, y_pred)
```

### 5 **Orden incorrecto de parámetros en accuracy_score**
```python
# El orden debería ser (y_true, y_pred), lo tienes invertido en una celda
metrics = {"Accuracy Train: ": accuracy_score(best_y_pred_train, y_train),  # 
           ...}

#  Correcto:
metrics = {"Accuracy Train: ": accuracy_score(y_train, best_y_pred_train),
           ...}
```

---

Buen trabajo Elius. Está bastante completo y bien estructurado. Los puntos a corregir:

1. **Error de variable**: `svc` no está definido en RandomizedSearchCV
2. **Guarda el vectorizer** - es imprescindible para predicciones futuras
3. **Añade classification_report** para ver precision, recall, f1

El preprocesamiento de texto está bien hecho y la estructura general es correcta. ¡Buen trabajo!

**Comment by elius123ef** [comment]
Corregidas las issues
